### ISSUE TITLE
HikariCP 5 - Connection is not available, request timed out after 300001ms.

### ISSUE THREAD

**Issue by KarthikeyanB007** [brettwooldridge/HikariCP#1817]
Hi,

I still see the issue with HikariCP V5 + OpenJDK 11. Could you please help with a resolution at the earliest possible? Thank you.

===

minimumIdle: 5
maximumPoolSize: 5
maxLifetime: 1800000
leakDetectionThreshold: 45000
keepaliveTime: 120000
connectionTimeout: 300000

===

When total becomes less than maximumPoolSize, why are nt new connections added to the pool to retain minimumIdle (which is = maximumPoolSize = 5)?

===

...
2021-07-26 13:15:13.321 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : Driver class org.postgresql.Driver found in Thread context class loader org.springframework.boot.loader.LaunchedURLClassLoader@68be2bc2
2021-07-26 13:15:13.442 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : prod-postingestion-pool - configuration:
2021-07-26 13:15:13.453 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : allowPoolSuspension.............false
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : autoCommit......................true
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : catalog.........................none
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : connectionInitSql...............none
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : connectionTestQuery.............none
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : connectionTimeout...............300000
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : dataSource......................none
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : dataSourceClassName.............none
2021-07-26 13:15:13.454 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : dataSourceJNDI..................none
2021-07-26 13:15:13.455 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : dataSourceProperties............{password=<masked>}
2021-07-26 13:15:13.457 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : driverClassName................."org.postgresql.Driver"
2021-07-26 13:15:13.457 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : exceptionOverrideClassName......none
2021-07-26 13:15:13.457 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : healthCheckProperties...........{}
2021-07-26 13:15:13.457 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : healthCheckRegistry.............none
2021-07-26 13:15:13.457 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : idleTimeout.....................600000
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : initializationFailTimeout.......1
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : isolateInternalQueries..........false
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : jdbcUrl.........................jdbc:postgresql://***:5432/***?currentSchema=***&stringtype=unspecified&sslmode=allow
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : keepaliveTime...................120000
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : leakDetectionThreshold..........45000
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : maxLifetime.....................1800000
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : maximumPoolSize.................5
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : metricRegistry..................none
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : metricsTrackerFactory...........none
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : minimumIdle.....................5
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : password........................<masked>
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : poolName........................"prod-postingestion-pool"
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : readOnly........................false
2021-07-26 13:15:13.458 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : registerMbeans..................false
2021-07-26 13:15:13.459 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : scheduledExecutor...............none
2021-07-26 13:15:13.459 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : schema..........................none
2021-07-26 13:15:13.459 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : threadFactory...................internal
2021-07-26 13:15:13.459 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : transactionIsolation............default
2021-07-26 13:15:13.459 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : username........................"***"
2021-07-26 13:15:13.459 DEBUG 97145 --- [           main] com.zaxxer.hikari.HikariConfig           : validationTimeout...............5000
2021-07-26 13:15:13.459  INFO 97145 --- [           main] com.zaxxer.hikari.HikariDataSource       : prod-postingestion-pool - Starting...
2021-07-26 13:15:14.289  INFO 97145 --- [           main] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Added connection org.postgresql.jdbc.PgConnection@733037
2021-07-26 13:15:14.291  INFO 97145 --- [           main] com.zaxxer.hikari.HikariDataSource       : prod-postingestion-pool - Start completed.
2021-07-26 13:15:14.348  INFO 97145 --- [           main] o.hibernate.jpa.internal.util.LogHelper  : HHH000204: Processing PersistenceUnitInfo [name: default]
2021-07-26 13:15:14.391 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=1, active=0, idle=1, waiting=0)
...
2021-07-26 21:41:47.679 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=5, active=0, idle=5, waiting=0)
2021-07-26 21:41:47.679 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Fill pool skipped, pool has sufficient level or currently being filled (queueDepth=0.
...
2021-07-26 21:42:06.578 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@2b18cf39 is alive
2021-07-26 21:42:17.679 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=5, active=0, idle=5, waiting=0)
2021-07-26 21:42:17.680 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Fill pool skipped, pool has sufficient level or currently being filled (queueDepth=0.
...
2021-07-26 21:42:40.350  WARN 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Failed to validate connection org.postgresql.jdbc.PgConnection@5f05f5ef (This connection has been closed.). Possibly consider using a shorter maxLifetime value.
2021-07-26 21:42:40.351 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Add connection elided, waiting=0, adders pending/running=1
2021-07-26 21:42:40.351 DEBUG 97145 --- [nnection closer] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Closing connection org.postgresql.jdbc.PgConnection@5f05f5ef: (connection is dead)
2021-07-26 21:42:40.351 DEBUG 97145 --- [nnection closer] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Closing connection org.postgresql.jdbc.PgConnection@2b18cf39: (connection has passed maxLifetime)

2021-07-27 14:33:18.262 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=5, active=0, idle=5, waiting=0)
2021-07-27 14:33:18.262 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Fill pool skipped, pool has sufficient level or currently being filled (queueDepth=0.
...
2021-07-27 14:33:34.807 DEBUG 97145 --- [nnection closer] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Closing connection org.postgresql.jdbc.PgConnection@3b2a1c76: (connection has passed maxLifetime)
2021-07-27 14:33:48.263 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=4, active=0, idle=4, waiting=0)
2021-07-27 14:34:12.700 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@6d16cbcd is alive
2021-07-27 14:34:18.263 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=4, active=0, idle=4, waiting=0)
...
2021-07-27 14:34:48.263 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=4, active=0, idle=4, waiting=0)
...
2021-07-27 14:34:54.699 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@6b4ef4a7 is alive
2021-07-27 14:35:04.665 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@763217ad is alive
2021-07-27 14:35:10.616 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@52857c47 is alive
2021-07-27 14:35:18.263 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=4, active=0, idle=4, waiting=0)
...
2021-07-27 14:35:48.263 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=4, active=0, idle=4, waiting=0)
2021-07-27 14:36:09.980 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@6d16cbcd is alive
2021-07-27 14:36:18.264 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=4, active=0, idle=4, waiting=0)
2021-07-27 14:36:44.154 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Add connection elided, waiting=0, adders pending/running=8
2021-07-27 14:36:44.154 DEBUG 97145 --- [nnection closer] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Closing connection org.postgresql.jdbc.PgConnection@763217ad: (connection has passed maxLifetime)
2021-07-27 14:36:46.545 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@6b4ef4a7 is alive
2021-07-27 14:36:48.264 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:37:08.340 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@52857c47 is alive
2021-07-27 14:37:18.264 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:37:48.264 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
2021-07-27 14:38:07.259 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@6d16cbcd is alive
2021-07-27 14:38:18.265 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:38:38.391 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@6b4ef4a7 is alive
2021-07-27 14:38:48.265 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:39:06.063 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@52857c47 is alive
2021-07-27 14:39:18.265 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:39:48.266 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:58:43.303 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@52857c47 is alive
2021-07-27 14:58:48.275 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:59:08.699 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - keepalive: connection org.postgresql.jdbc.PgConnection@6b4ef4a7 is alive
2021-07-27 14:59:18.276 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=3, active=0, idle=3, waiting=0)
...
2021-07-27 14:59:35.641 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Add connection elided, waiting=0, adders pending/running=17
2021-07-27 14:59:35.641 DEBUG 97145 --- [nnection closer] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Closing connection org.postgresql.jdbc.PgConnection@6d16cbcd: (connection has passed maxLifetime)
2021-07-27 14:59:48.276 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=2, active=0, idle=2, waiting=0)
...
2021-07-27 15:02:48.277 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=2, active=0, idle=2, waiting=0)
...
2021-07-27 15:02:51.934 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Add connection elided, waiting=0, adders pending/running=17
2021-07-27 15:02:51.935 DEBUG 97145 --- [nnection closer] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Closing connection org.postgresql.jdbc.PgConnection@6b4ef4a7: (connection has passed maxLifetime)
2021-07-27 15:03:12.268 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Add connection elided, waiting=0, adders pending/running=17
2021-07-27 15:03:12.268 DEBUG 97145 --- [nnection closer] com.zaxxer.hikari.pool.PoolBase          : prod-postingestion-pool - Closing connection org.postgresql.jdbc.PgConnection@52857c47: (connection has passed maxLifetime)
2021-07-27 15:03:18.279 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=0)
...
2021-07-27 15:03:48.279 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=0)
2021-07-27 15:04:18.279 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=0)
...
2021-07-27 15:04:48.279 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=0)
...
2021-07-27 15:04:51.612 DEBUG 97145 --- [   scheduling-1] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Add connection elided, waiting=1, adders pending/running=17
2021-07-27 15:05:18.280 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=1)
2021-07-27 15:05:48.280 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=1)
...
2021-07-27 15:09:18.282 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=1)
2021-07-27 15:09:48.282 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Pool stats (total=0, active=0, idle=0, waiting=1)
2021-07-27 15:09:51.612 DEBUG 97145 --- [   scheduling-1] com.zaxxer.hikari.pool.HikariPool        : prod-postingestion-pool - Timeout failure stats (total=0, active=0, idle=0, waiting=0)
2021-07-27 15:09:51.622  WARN 97145 --- [   scheduling-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: null
2021-07-27 15:09:51.622 ERROR 97145 --- [   scheduling-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : prod-postingestion-pool - Connection is not available, request timed out after 300001ms.
2021-07-27 15:09:51.631 ERROR 97145 --- [   scheduling-1] o.s.s.s.TaskUtils$LoggingErrorHandler    : Unexpected error occurred in scheduled task

org.springframework.dao.DataAccessResourceFailureException: Unable to acquire JDBC Connection; nested exception is org.hibernate.exception.JDBCConnectionException: Unable to acquire JDBC Connection
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:277) ~[spring-orm-5.2.6.RELEASE.jar!/:5.2.6.RELEASE]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:255) ~[spring-orm-5.2.6.RELEASE.jar!/:5.2.6.RELEASE]
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:528) ~[

===

**Comment by brettwooldridge** [comment]
This log...
```
2021-07-27 14:59:35.641 DEBUG 97145 --- [ool housekeeper] com.zaxxer.hikari.pool.HikariPool : prod-postingestion-pool - Add connection elided, waiting=0, adders pending/running=17
```
indicates that the connection adder thread is running ... and there are 17 other "adder tasks" piled up behind that one, waiting to add.

This means that the "adder" thread is simply "stuck" trying to add a connection. It is calling ``getConnection()`` on the underlying driver, and then timing out or failing, and then retrying.

The "adder task" loops _infinitely_ until it can fill the pool upto at least ``minimumIdle`` connections.

I recommend forcing a full stacktrace dump to check where the adder task thread is waiting -- or viewing the thread stacks through a JMX console.  My guess is the adder task is parked in either ``getConnection()`` on the underlying DataSource, or in the sleep before retry. If the ``getConnection()`` call is failing instantly on the underlying DataSource, you will see the adder task spending most of its time in the sleep between retries. If the ``getConnection()`` on the underlying datasource is timing out (not failing instantly) then you will likely see the thread blocked on that method.

**Comment by greenwoodct** [comment]
This issue is related to recently added new issue "[Not removing a dead connection from the pool 1819](https://github.com/brettwooldridge/HikariCP/issues/1819)".

**Comment by KarthikeyanB007** [comment]
"I recommend forcing a full stacktrace dump" -> I use PostGres driver. Is there anything I need to do to enable this? I have done all in my powers to extract maximum logs.

**Comment by brettwooldridge** [comment]
@KarthikeyanB007 Here are several ways to get stacktrace dumps:

https://access.redhat.com/solutions/18178

**Comment by rrockx-trifork** [comment]
We are experiencing similar problems. We did not have the logging enabled, instead we read out the pool stats using Spring Actuator. However, we did create a thread dump at the time, which might be helpful:

```
"HikariPool-1 connection adder" #193102 daemon prio=5 os_prio=0 cpu=7.29ms elapsed=3499.00s tid=0x000055b81eebd800 nid=0x7892 runnable  [0x00007fa3a3f89000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(java.base@11.0.5/Native Method)
	at java.net.SocketInputStream.socketRead(java.base@11.0.5/SocketInputStream.java:115)
	at java.net.SocketInputStream.read(java.base@11.0.5/SocketInputStream.java:168)
	at java.net.SocketInputStream.read(java.base@11.0.5/SocketInputStream.java:140)
	at sun.security.ssl.SSLSocketInputRecord.read(java.base@11.0.5/SSLSocketInputRecord.java:448)
	at sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(java.base@11.0.5/SSLSocketInputRecord.java:68)
	at sun.security.ssl.SSLSocketImpl.readApplicationRecord(java.base@11.0.5/SSLSocketImpl.java:1104)
	at sun.security.ssl.SSLSocketImpl$AppInputStream.read(java.base@11.0.5/SSLSocketImpl.java:823)
	- locked <0x00000000f6c22358> (a sun.security.ssl.SSLSocketImpl$AppInputStream)
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:161)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:128)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:113)
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:73)
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:370)
	at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:510)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:146)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:197)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:211)
	at org.postgresql.Driver.makeConnection(Driver.java:459)
	at org.postgresql.Driver.connect(Driver.java:261)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:358)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:477)
	at com.zaxxer.hikari.pool.HikariPool.access$100(HikariPool.java:71)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:725)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:711)
	at java.util.concurrent.FutureTask.run(java.base@11.0.5/FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.5/ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.5/ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(java.base@11.0.5/Thread.java:834)
```

So in our case there were also no connections available anymore and no connections were added either.
Looking at this part of the thread dump, it looks like PostgreSQL JDBC driver was able to setup a connection, but got stuck while authenticating.
We did not configure socket timeouts for the data source, so I assume the connection became stale and this process was waiting infinite for the read to end. So setting a data source socket timeout might solve this, but, not sure I am correct, this will limit the execution time of your queries as well.
Another way to solve this is by setting the property _loginTimeout_ of the PostgreSQL JDBC driver, which will only timeout the login process and not the execution of the query as a whole (for which we use a Spring transaction timeout).

Just like mentioned here https://github.com/brettwooldridge/HikariCP/issues/1819#issuecomment-909875464 both the client and the PostgreSQL server are hosted on Azure.

**Comment by PedroMPagani** [comment]
I've also been experiencing this issue, and I was wondering, should I be using autoReconnect=true on jdbc url?

**Comment by toniclecb** [comment]
I've also been experiencing this issue.
I've already changed the properties autoReconnect=true to false and setMaximumPoolSize to higher and lower numbers

Here is my stacktrace:
```
[31merror[0m] from application in application-akka.actor.default-dispatcher-3 - 
java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30001ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:181)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:146)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
```

```
from com.zaxxer.hikari.pool.PoolBase in xx - 5 connection closer - xx - 5 - Closing connection com.mysql.jdbc.JDBC4Connection@399823e1: (connection has passed maxLifetime)
from com.zaxxer.hikari.pool.HikariPool in xx - 5 connection adder - xx - 5 - Added connection com.mysql.jdbc.JDBC4Connection@29681660
from com.zaxxer.hikari.pool.PoolBase in xx - 3 connection closer - xx - 3 - Closing connection com.mysql.jdbc.JDBC4Connection@292be283: (connection has passed maxLifetime)
from com.zaxxer.hikari.pool.HikariPool in xx - 3 connection adder - xx - 3 - Added connection com.mysql.jdbc.JDBC4Connection@1d785115
from com.zaxxer.hikari.pool.PoolBase in xx - 5 connection closer - xx - 5 - Closing connection com.mysql.jdbc.JDBC4Connection@646e1c68: (connection has passed maxLifetime)
from com.zaxxer.hikari.pool.HikariPool in xx - 5 connection adder - xx - 5 - Added connection com.mysql.jdbc.JDBC4Connection@361f2f1
from com.zaxxer.hikari.pool.HikariPool in application-akka.actor.default-dispatcher-5 - HikariPool-1 - Timeout failure stats (total=64, active=64, idle=0, waiting=10)
from com.zaxxer.hikari.pool.HikariPool in application-akka.actor.default-dispatcher-6 - HikariPool-1 - Timeout failure stats (total=64, active=64, idle=0, waiting=9)
from com.zaxxer.hikari.pool.HikariPool in application-akka.actor.default-dispatcher-3 - HikariPool-1 - Timeout failure stats (total=64, active=64, idle=0, waiting=8)
```

It's seems to freeze when calling getConnection().
![image](https://user-images.githubusercontent.com/1041102/154564730-890dd313-c500-4221-ad61-789bfc999710.png)

Yesterday I deployed the version 5.0.1, today I'm already getting this error.
Is there something I can do here to make this stop?

**Comment by lfbayer** [comment]
@toniclecb, Your situation sounds different, and much simpler. The log says you have 64 "active" connections. This means that there have been 64 connections borrowed from the pool that were never returned. You need to make sure that you close connections after you use them in order to return them to the pool.

**Comment by toniclecb** [comment]
Thanks @lfbayer 
I've already define setLeakDetectionThreshold().
Today I got this:
```
3 housekeeper - Connection leak detection triggered for com.mysql.jdbc.JDBC4Connection@753bed27 on thread pool-6-thread-4, stack trace follows
3 housekeeper - Connection leak detection triggered for com.mysql.jdbc.JDBC4Connection@554264a5 on thread pool-6-thread-4, stack trace follows

thread-7 - Previously reported leaked connection com.mysql.jdbc.JDBC4Connection@1b4967c3 on thread pool-6-thread-4 was returned to the pool (unleaked)
thread-9 - Previously reported leaked connection com.mysql.jdbc.JDBC4Connection@5569decd on thread pool-6-thread-4 was returned to the pool (unleaked)
```
This means the two connections passed the time defined in setLeakDetectionThreshold, but were closed after that?




The stats now are similar to this:
```
1 housekeeper - AMTI - 1 - Pool stats (total=24, active=2, idle=22, waiting=0)
2 housekeeper - INATIVOS - Pool stats (total=24, active=0, idle=24, waiting=0)
4 housekeeper - AMTI - 4 - Pool stats (total=24, active=3, idle=21, waiting=0)
2 housekeeper - AMTI - 2 - Pool stats (total=24, active=0, idle=24, waiting=0)
```

**Comment by PedroMPagani** [comment]
Toni, there might be an easy solution for you which is use the DataSourceUtil class, I was having the same exceptions until I just ended up using it, 

`
 return DataSourceUtils.getConnection(databasePool.getDataSource());
`

`
        <dependency>
            <groupId>org.springframework</groupId>
            <artifactId>spring-jdbc</artifactId>
            <version>5.3.15</version>
            <scope>compile</scope>
        </dependency>
`

When i had that exception all the reports i saw most of them said it would fix it.

**Comment by brettwooldridge** [comment]
@PedroMPagani This is good advice. The doc for ``DataSourceUtils.getConnection()`` says:

> Is aware of a corresponding Connection bound to the current thread, for example when using [DataSourceTransactionManager](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/datasource/DataSourceTransactionManager.html). Will bind a Connection to the thread if transaction synchronization is active, e.g. when running within a [JTA](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/jta/JtaTransactionManager.html) transaction).

My company's application implements a similar class to ``DataSourceUtils`` that returns the same Connection that a thread already holds if it requests another Connection, instead of consuming another one from the pool.

**Comment by PedroMPagani** [comment]
That's right @brettwooldridge, I didn't mention the doc trusting that fellow users would research the reason why, thank you!
